### PR TITLE
WASM and AudioWorklet improvements

### PR DESCRIPTION
**Pull request by chrisuehlinger**
I've checked in what I've got so far. Here are some thoughts and questions I have at this point.

Thoughts:
- In my first commit I fix the addFunction signature bug that cropped up in emcc v1.38.48, the malloc bug that cropped up in v2.0.15, and the issues with AutoplayPolicy and Web Audio Contexts. If nothing else, you definitely want that commit.
- The AudioWorklet stuff works, but could use a bit of cleanup (I've got bad/debug flags and logging in some places). If you like what I've got so far I can go all the way and polish it up.
- Emscripten isn't great at supporting Worklets yet (they're a little different from Workers, they don't have `self` or `importScripts`) but with a couple small hacks (see templates/hv_worklet_start.js) we can trick it reliably, which could do until an `-sENVIRONMENT=worklet` option gets added in the future.
- The Worklet needs to call `registerProcessor()` in its starting stackframe, as after that `await this.webAudioContext.audioWorklet.addModule()` resolves and if the processor isn't registered then `new AudioWorkletNode()` will throw. To make that work I compile the Worklet WASM with `-sMODULARIZE=0` and `-sSINGLE_FILE=1` so that we don't have to wait for a fetch call or even an immediate Promise resolution.
- Allegedly `-sSINGLE_FILE=1` used to just dump asm.js into the file instead of a base64-encoded WASM string. Emscripten fixed that a while back, so it's safe and shouldn't incur any performance penalty. That's why in my third commit (the most brazen one) I switch to compiling all WASM that way.

Questions:
- Should we remove the link to Enzien Audio's website from index.html? It looks like they've lost control of their domain.
- I've wired up the setFloatParameter method to pass its value to the Worklet. Do you have an example of patches that use some of those other functions? I seem to recall reading that hvcc doesn't support bangs, only loadbangs.
- In my local checkout I have a Dockerfile/.dockerignore/etc setup that I've been using to test this (so I don't have to manage all the emcc stuff on my personal drive). Would you like me to check it in? It might help other people get up and running.

### CODE REVIEW

**Comment by dromer** [comment]
Did a quick test of your branch, using emcc 3.1.3 (that I still had installed. will try latest in a bit) and the result works for me on chromium, but not on firefox.

It might be good to keep a fallback to older methods for this, but that might also mean some extra work.

rebeltech.org maintains an online compile + js hosting for patches that works on all the major browsers. ideally we try to match that ;)

**Comment by chrisuehlinger** [comment]
> Did a quick test of your branch, using emcc 3.1.3 (that I still had installed. will try latest in a bit) and the result works for me on chromium, but not on firefox.

OK, I'll look into this now.

> rebeltech.org maintains an online compiler + js hosting for patches that works on all the major browsers. ideally we try to match that ;)

That works, I'll look up what they support.

> yeah, lets just point to this repository, would be more "stable" of a url

Sounds good.

> I've added the ability to mark parameters as a type which can then be used in the wrapper to treat them differently. Ie. for DPF wrapper I use the bool and trig types to append parameter hints. Of course inside the patch you still have to make the logic to create a bang based on such a trigger (which should then only respond to a 1 value). Bit of a workaround, but usable.

OK cool, I'll look into this.

> I'm not sure what your dockerfile would do, preferably we don't put such specific dockerfiles, they wouldn't be useful for any other generator so I don't see the purpose. IMO the idea of this js generator is that people should have emcc in the path. How you get it there is the responsibility of whoever uses it. I think it would be good to mark which emsdk/emcc versions we can support using the functioning wrapper, though.

Makes sense, I'll keep it local. I started using it when I was trying to get a working version of emcc 1.37.38, since I kept having to install and remove various libraries and other junk. Now that we can use whatever modern emcc is in the PATH, it's a lot less needed.

**Comment by chrisuehlinger** [comment]
The Firefox fix turned out to be easier than I expected: We just need to compile/instantiate the WASM synchronously when in the Worklet. I'm guessing the issue came down to Chrome and Firefox's runtimes differing in how they schedule resolved Promises in the event loop.

**Comment by dromer** [comment]
Cool, totally works now!

If we really want feature parity with rebeltech we should also add webmidi (chrome only), but that is maybe a real enhancement we could look at for the future.

File: hvcc/generators/c2js/c2js.py [hunk 1]
```diff
@@ -185,6 +190,39 @@ def compile(clazz, c_src_dir, out_dir, externs,
...
+                f.write(env.get_template("hv_worklet_start.js").render(
+                    name=patch_name,
+                    copyright=copyright_js,
+                    externs=externs,
+                    pool_sizes_kb=externs["memoryPoolSizesKb"]))
+
+            js_path = c2js.run_emscripten(c_src_dir=c_src_dir,
+                                          out_dir=out_dir,
+                                          patch_name=patch_name,
+                                          output_name=patch_name+"_AudioLibWorklet",
```
Comment (dromer) [inline hvcc/generators/c2js/c2js.py:215]: please use f-strings for combining variables and strings. so this would be `f"{patch_name}_AudioLibWorklet"`

**Comment by CarloCattano** [comment]
Amazing job done here! thanks a lot 
Im trying the pull requesto on windows, just installed latest emsdk and its on my PATH so it runs normally:
``` emcc
emcc: error: no input files
```
But hvcc seems to be hardcoded to search for it on a specific folder with a specific version , is this intentional ? 
would be great to have it accept whatever its on PATH : 
```
dev\hvcc\examples> hvcc .\example.pd -o wasmtest -n wasmTEST -g js
  1) Error c2js: Emscripten emcc command line tool not installed, should be here: c:\users\freev\downloads\dev\hvcc\hvcc\generators\c2js\emsdk\emscripten\1.37.38\emcc
```

**Comment by CarloCattano** [comment]
```--> Generating Javascript
  1) Error c2js: Emscripten emcc command line tool not installed, should be here: c:\users\freev\downloads\dev\hvcc\hvcc\generators\c2js\emsdk\emscripten\1.37.38\emcc
Total compile time: 215.01ms```

On windows I have to manually set the PATH on c2js.py as its installed by reading default docs of emsdk to :
```
 emcc_path = os.path.abspath(os.path.join(
            os.path.dirname(__file__), "emsdk", "upstream","emscripten", "emcc.bat"))

        emcpp_path = os.path.abspath(os.path.join(
            os.path.dirname(__file__), "emsdk", "upstream","emscripten", "em++.bat"))
```
If I dont add the .bat I get an error ```c2js: [WinError 193] %1 is not a valid Win32 application

and even then , It goes through till it gets an error at : 
```
--> Generating Javascript
In file included from wasmtest\c\HvLightPipe.c:17:
In file included from wasmtest\c/HvLightPipe.h:20:
wasmtest\c/HvUtils.h:39:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
emcc: error: 'C:/Users/x/Downloads/dev/hvcc/hvcc/generators/c2js/emsdk/upstream/bin\clang.exe -target wasm32-unknown-emscripten -DEMSCRIPTEN -D__EMSCRIPTEN_major__=3 -D__EMSCRIPTEN_minor__=1 -D__EMSCRIPTEN_tiny__=14 -fignore-exceptions -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=C:\Users\x\Downloads\dev\hvcc\hvcc\generators\c2js\emsdk\upstream\emscripten\cache\sysroot -Xclang -iwithsysroot/include\compat -Iwasmtest\c -DHV_SIMD_NONE -ffast-math -DNDEBUG -Wall -c wasmtest\c\HvLightPipe.c -o wasmtest\c\HvLightPipe.o' failed (returned 1)
  1) Error c2js: Command '['c:\\users\\x\\downloads\\dev\\hvcc\\hvcc\\generators\\c2js\\emsdk\\upstream\\emscripten\\emcc.bat', '-Iwasmtest\\c', '-DHV_SIMD_NONE', '-ffast-math', '-DNDEBUG', '-Wall', '-c', '-o', 'wasmtest\\c\\HvLightPipe.o', 'wasmtest\\c\\HvLightPipe.c']' returned non-zero exit status 1.
Total compile time: 540.84ms ```

So from reading the PR I agree that it would be the best to just find the emcc anywhere instead of on a fixed PATH with version number etc.

And about this error I have not clue , need more time to figure it out. Will test on linux to try it at last
Thanks a lot for your efforts bringing it back to live !

**Comment by chrisuehlinger** [comment]
@CarloCattano Could you upload the patch and the output folder you're using to test? That warning will come up if the audio worklet throws an exception while initialzing.

**Comment by CarloCattano** [comment]
I see ! Might be cause its an old patch I used before in heavy using [heavylib ](https://github.com/Wasted-Audio/heavylib) then 
[fmTest.zip](https://github.com/Wasted-Audio/hvcc/files/8978819/fmTest.zip)
Its just an hv.filter , not really needed but hope it can give you some insight for development
Should I bug you with issues on your repo instead of spamming here the PR ?

Still could not make it work on windows 

thanks @chrisuehlinger

**Comment by CarloCattano** [comment]
@chrisuehlinger issues are disabled I think

**Comment by chrisuehlinger** [comment]
> @chrisuehlinger issues are disabled I think

Should be good to go now, I didn't realize they were opt-in.

File: hvcc/generators/c2js/c2js.py (hunk 1 above)
Comment (chrisuehlinger) [reply hvcc/generators/c2js/c2js.py]: Should be good on this now

**Comment by chrisuehlinger** [comment]
So I was doing some cross-browser testing, thinking that if the issues the two of you were having were just environment issues or server/MIME issues then we could fix those and I'd have a better case for arguing for dropping the ScriptNodeProcessor.

But then I opened Safari, it wasn't working, I did some googling, and found this. I have no idea what the WebKit team is doing here, but apparently they claim that Safari "supports" AudioWorklets, but "doesn't support" the `AudioWorklet.addModule()` function that you need in order to load them (and elsewhere it says that they don't support the `AudioWorkletNode()` constructor either).

Screen Shot 2022-06-30 at 11 31 46 PM

Unfortunately, in Safari `!!audioContext.audioWorklet.addModule` actually evaluates to true, and calling the function doesn't throw an exception, so I'll need to find a craftier way to feature-detect this. Until this is resolved though, it'll block merging, since we don't want to just do nothing on Safari.

**Comment by chrisuehlinger** [comment]
OK, so I retract what I said before: The compat table is wrong and Safari does in fact support the full AudioWorklet spec including `.addModule()` and the `AudioWorkletNode()` constructor (as evidence, [this](https://googlechromelabs.github.io/web-audio-samples/audio-worklet/basic/hello-audio-worklet/) works in Safari). However, Safari will not display anything you log to the console in an AudioWorklet and doesn't support the `onprocesserror` event, which together make debugging difficult.

Thus far, I've cloned the ChromeLabs repo with that hello world page and tried to get it to run one of our Worklets. I started with their code and copied things over until it broke, then took notes. It looks like setting `this.sampleRate` in an AudioWorklet in Safari breaks it, as does setting the printHook and sendHook. Commenting those lines out, I was able to get everything else to not crash, but I couldn't get it to read from the processBuffer. This at least gives us something to work with, and I can follow up later.

Having worked with Safari and WebRTC a lot, I know that sometimes they wait for a standard to become frozen, then only implement the latest version of it and do it in a really strict way. So it's possible that the answers are in [the W3C Spec](https://www.w3.org/TR/webaudio/), which I'm poring over in another tab (I refuse to let Safari defeat me).

That's all for tonight, I may be able to another pass at this over the weekend if there's a rainy day.

**Comment by chrisuehlinger** [comment]
I took another crack, and I've got a build that works in Safari, I just need to port all my changes back to the templates. The issue was: Safari only lets you resume an `AudioContext` in response to a user gesture like a click. But unlike other browsers, Safari requires you to call `.resume()` **within the same stack-frame**, so if you try to do it inside of an `async` function it won't work because `async` functions run after the callstack has cleared.

To remedy this, I've added a `Promise` that starts after the `AudioContext` has been created (and can be resumed) however any parameters will have to be sent over after the promise resolves.

> Btw, I've been slowly thinking/struggling about organizing some example patches in a separate repo. One idea was to then have different CI for each of the generators based on that. (compared to the main hvcc testing all the specific generators may require quite some elaborate CI steps, making the whole build process many times longer as well. Would rather try and treat the generators as their own thing for this.
> 
> Anyway, my question was if you have an idea how such webasm builds could be tested, maybe? (trying to think about some long term stability goals for the project)
>
> [edit: maybe useful -> https://github.com/montag/audioworklet-test-demo can start a separate discussion thread or some "todo" on the project-page]

My guess is that the Web Audio stuff will be harder to test than the WASM. In my experience, trying to test against slightly non-deterministic outputs (like screenshots of a page or a waveform of audio) is very fragile and unreliable. However, given what we're doing here, I think the best way to test this would be to use something like [Puppeteer](https://github.com/puppeteer/puppeteer) to run the example patch pages and just make sure they don't have console errors or crashes. That alone will probably help a lot.

> Btw @chrisuehlinger I will add you to the repo. You seem to know infinitely more about webasm than me and would be happy to have you on board.

Thanks! I can't promise I'll always be as available as I am now (I'm currently between theatre shows, so I have spare time for the next month or so) but I'm down to help out where and when I can!

**Comment by chrisuehlinger** [comment]
One last thing: This update is definitely going to drop IE11 support. There are just too many browser features that are necessary for this modern audio workflow that won't work on IE11. The good(ish) news is that the Patch editor you referenced in the other thread also doesn't support IE11 (I checked using my Windows 10 partition, as IE is not available on Windows 11), so we are maintaining parity with them.

**Comment by dromer** [comment]
Would be good to include a table of "minimum supported browser versions" or something.

The fallback was nice (and still have loads of old devices that theoretically can still play webasm in some form), but maintaining all of that backwards compat is just not doable. Especially since we're all on hobby time with this kind of stuff :)

And no worries! It's more that I also want to open up this project to active community members and not just me making all the shots.

One thing I find odd is that the CI isn't running for your build. (normally it at least shows a manual trigger, for external collaborators)

**Comment by chrisuehlinger** [comment]
> Would be good to include a table of "minimum supported browser versions" or something.

So one tough thing with this is that it's hard to actually get your hands on old versions of some browsers and actually validate that they work. I think a good compromise would be to look at the features we're using (WebAssembly, ScriptProcessorNode||AudioWorklet, Promises, async/await, etc.), see what the least common denominator is in terms of support, and declare support for those browsers and versions. If someone is having an issue with a particular old browser version we can handle that when the time comes.

> The fallback was nice (and still have loads of old devices that theoretically can still play webasm in some form), but maintaining all of that backwards compat is just not doable. Especially since we're all on hobby time with this kind of stuff :)

For the record, the branch as-is retains the ScriptProcessorNode as a fallback. I'm not sure how often it'll get used, but having the WASM on the main thread also makes it easier to boot up the AudioWorklet because we can figure out how many output channels we need ahead of time (we can find a better way to do this in the future). The thing I've deprecated is asm.js.

> One thing I find odd is that the CI isn't running for your build. (normally it at least shows a manual trigger, for external collaborators)

Is it OK if I just merge this with `js_generator` and then make any last minute tweaks there?

**Comment by chrisuehlinger** [comment]
I started looking up caniuse tables for the features, and it turns out IE11 doesn't even support ScriptProcessorNode, so even hvcc 0.5.0 doesn't support it.

**Comment by dromer** [comment]
I left one final comment, but if you think it's good to merge then lets proceed!

(quarterly release is coming up and this would be good to include :) )

**Pull request merged**